File: appverse-core/src/csharp/Unity/Core/IO/ScriptSerialization/JSON/JsonSerializer.cs
/*
 Copyright (c) 2012 GFT Appverse, S.L., Sociedad Unipersonal.

 This Source  Code Form  is subject to the  terms of  the Appverse Public License 
 Version 2.0  (“APL v2.0”).  If a copy of  the APL  was not  distributed with this 
 file, You can obtain one at http://appverse.org/legal/appverse-license/.

 Redistribution and use in  source and binary forms, with or without modification, 
 are permitted provided that the  conditions  of the  AppVerse Public License v2.0 
 are met.

 THIS SOFTWARE IS PROVIDED BY THE  COPYRIGHT HOLDERS  AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS  OR IMPLIED WARRANTIES, INCLUDING, BUT  NOT LIMITED TO,   THE IMPLIED
 WARRANTIES   OF  MERCHANTABILITY   AND   FITNESS   FOR A PARTICULAR  PURPOSE  ARE
 DISCLAIMED. EXCEPT IN CASE OF WILLFUL MISCONDUCT OR GROSS NEGLIGENCE, IN NO EVENT
 SHALL THE  COPYRIGHT OWNER  OR  CONTRIBUTORS  BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL,  SPECIAL,   EXEMPLARY,  OR CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT
 LIMITED TO,  PROCUREMENT OF SUBSTITUTE  GOODS OR SERVICES;  LOSS OF USE, DATA, OR
 PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) 
 ARISING  IN  ANY WAY OUT  OF THE USE  OF THIS  SOFTWARE,  EVEN  IF ADVISED OF THE 
 POSSIBILITY OF SUCH DAMAGE.
 */
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Collections;
using System.Reflection;
using System.ComponentModel;

namespace Unity.Core.IO.ScriptSerialization.JSON
{
	/// <summary>
	/// Specifies reference loop handling options for the <see cref="JsonWriter"/>.
	/// </summary>
	enum ReferenceLoopHandling
	{
		/// <summary>
		/// Throw a <see cref="JsonSerializationException"/> when a loop is encountered.
		/// </summary>
		Error = 0,
		/// <summary>
		/// Ignore loop references and do not serialize.
		/// </summary>
		Ignore = 1,
		/// <summary>
		/// Serialize loop references.
		/// </summary>
		Serialize = 2
	}

	/// <summary>
	/// Serializes and deserializes objects into and from the Json format.
	/// The <see cref="JsonSerializer"/> enables you to control how objects are encoded into Json.
	/// </summary>
	sealed class JsonSerializer
	{
		sealed internal class DeserializerLazyDictionary : JavaScriptSerializer.LazyDictionary
		{
			readonly JsonReader _reader;
			readonly JsonSerializer _serializer;
			IEnumerator<KeyValuePair<string, object>> _innerEnum;
			object _firstValue;

			public DeserializerLazyDictionary (JsonReader reader, JsonSerializer serializer)
			{
				_reader = reader;
				_serializer = serializer;
			}

			public object PeekFirst ()
			{
				if (_innerEnum != null)
					throw new InvalidOperationException ("first already taken");

				_innerEnum = _serializer.PopulateObject (_reader);

				if (_innerEnum.MoveNext ())
					_firstValue = _innerEnum.Current;

				return _firstValue;
			}

			protected override IEnumerator<KeyValuePair<string, object>> GetEnumerator ()
			{
				if (_innerEnum == null)
					_innerEnum = _serializer.PopulateObject (_reader);

				if (_firstValue != null)
					yield return (KeyValuePair<string, object>)_firstValue;

				while (_innerEnum.MoveNext ())
					yield return _innerEnum.Current;
			}
		}

		sealed class SerializerLazyDictionary : JavaScriptSerializer.LazyDictionary
		{
			readonly object _source;

			public SerializerLazyDictionary (object source)
			{
				_source = source;
			}

			protected override IEnumerator<KeyValuePair<string, object>> GetEnumerator ()
			{
				foreach (MemberInfo member in ReflectionUtils.GetFieldsAndProperties (_source.GetType (), BindingFlags.Public | BindingFlags.Instance)) {
					if (ReflectionUtils.CanReadMemberValue (member) && !member.IsDefined (typeof(ScriptIgnoreAttribute), true))
					if (!ReflectionUtils.IsIndexedProperty (member))
						yield return new KeyValuePair<string, object> (member.Name, ReflectionUtils.GetMemberValue (member, _source));
				}
			}
		}

		sealed class GenericDictionaryLazyDictionary : JavaScriptSerializer.LazyDictionary
		{
			readonly object _source;
			readonly PropertyInfo _piKeys;
			readonly PropertyInfo _piValues;

			public GenericDictionaryLazyDictionary (object source, Type dictType)
			{
				_source = source;
				_piKeys = dictType.GetProperty ("Keys");
				_piValues = dictType.GetProperty ("Values");
			}

			protected override IEnumerator<KeyValuePair<string, object>> GetEnumerator ()
			{
				
				IEnumerable eKeys = (IEnumerable)_piKeys.GetValue (_source, null);
				IEnumerator eValues = ((IEnumerable)_piValues.GetValue (_source, null)).GetEnumerator ();
				foreach (object key in eKeys) {
					string keyString = key == null ? null : key.ToString ();
					if (!eValues.MoveNext ())
						throw new IndexOutOfRangeException (keyString);


					yield return new KeyValuePair<string, object> (keyString, eValues.Current);
				}

				if (eValues.MoveNext ())
					throw new IndexOutOfRangeException (eValues.Current != null ? eValues.Current.ToString () : String.Empty);
			}
		}

		private int _maxJsonLength;
		private int _recursionLimit;
		private int _currentRecursionCounter;
		private ReferenceLoopHandling _referenceLoopHandling;
		readonly JavaScriptSerializer _context;
		readonly JavaScriptTypeResolver _typeResolver;

		public int MaxJsonLength {
			get { return _maxJsonLength; }
			set { _maxJsonLength = value; }
		}

		public int RecursionLimit {
			get { return _recursionLimit; }
			set { _recursionLimit = value; }
		}

		/// <summary>
		/// Get or set how reference loops (e.g. a class referencing itself) is handled.
		/// </summary>
		public ReferenceLoopHandling ReferenceLoopHandling {
			get { return _referenceLoopHandling; }
			set {
				if (value < ReferenceLoopHandling.Error || value > ReferenceLoopHandling.Serialize) {
					throw new ArgumentOutOfRangeException ("value");
				}
				_referenceLoopHandling = value;
			}
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="JsonSerializer"/> class.
		/// </summary>
		public JsonSerializer (JavaScriptSerializer context, JavaScriptTypeResolver resolver)
		{
			_context = context;
			_typeResolver = resolver;
			_referenceLoopHandling = ReferenceLoopHandling.Error;
		}

		#region Deserialize
		public object Deserialize (TextReader reader)
		{
			return Deserialize (new JsonReader (reader, MaxJsonLength, RecursionLimit));
		}

		/// <summary>
		/// Deserializes the Json structure contained by the specified <see cref="JsonReader"/>
		/// into an instance of the specified type.
		/// </summary>
		/// <param name="reader">The type of object to create.</param>
		/// <param name="objectType">The <see cref="Type"/> of object being deserialized.</param>
		/// <returns>The instance of <paramref name="objectType"/> being deserialized.</returns>
		object Deserialize (JsonReader reader)
		{
			if (!reader.Read ())
				return null;

			return GetObject (reader);
		}

		private object GetObject (JsonReader reader/*, Type objectType*/)
		{
			if (RecursionLimit > 0 && reader.CurrentRecursionLevel >= RecursionLimit) {
				throw new ArgumentException ("RecursionLimit exceeded.");
			}

			object value;

			switch (reader.TokenType) {
			// populate a typed object or generic dictionary/array
			// depending upon whether an objectType was supplied
			case JsonToken.StartObject:
				//value = PopulateObject(reader/*, objectType*/);
				value = new DeserializerLazyDictionary (reader, this);
				break;
			case JsonToken.StartArray:
				value = PopulateList (reader/*, objectType*/);
				break;
			case JsonToken.Integer:
			case JsonToken.Float:
			case JsonToken.String:
			case JsonToken.Boolean:
			case JsonToken.Date:
				//value = EnsureType(reader.Value, objectType);
				value = reader.Value;
				break;
			case JsonToken.Constructor:
				value = reader.Value.ToString ();
				break;
			case JsonToken.Null:
			case JsonToken.Undefined:
				value = null;
				break;
			default:
				throw new JsonSerializationException ("Unexpected token whil deserializing object: " + reader.TokenType);
			}

			return value;
		}

		private IEnumerable<object> PopulateList (JsonReader reader/*, Type objectType*/)
		{

			while (reader.Read()) {
				switch (reader.TokenType) {
				case JsonToken.EndArray:
					yield break;
				case JsonToken.Comment:
					break;
				default:
					yield return GetObject (reader/*, elementType*/);

					break;
				}
			}

			throw new JsonSerializationException ("Unexpected end when deserializing array.");
		}

		private IEnumerator<KeyValuePair<string, object>> PopulateObject (JsonReader reader/*, Type objectType*/)
		{
			reader.IncrementRecursionLevel ();
			while (reader.Read ()) {
				switch (reader.TokenType) {
				case JsonToken.PropertyName:
					string memberName = reader.Value.ToString ();

					if (!reader.Read ())
						throw new JsonSerializationException (string.Format ("Unexpected end when setting {0}'s value.", memberName));
					yield return new KeyValuePair<string, object> (memberName, GetObject (reader));
					break;
				case JsonToken.EndObject:
					reader.DecrementRecursionLevel ();
					yield break;
				default:
					throw new JsonSerializationException ("Unexpected token when deserializing object: " + reader.TokenType);
				}
			}

			throw new JsonSerializationException ("Unexpected end when deserializing object.");
		}
		#endregion

		#region Serialize
		/// <summary>
		/// Serializes the specified <see cref="Object"/> and writes the Json structure
		/// to a <c>Stream</c> using the specified <see cref="TextWriter"/>. 
		/// </summary>
		/// <param name="textWriter">The <see cref="TextWriter"/> used to write the Json structure.</param>
		/// <param name="value">The <see cref="Object"/> to serialize.</param>
		public void Serialize (TextWriter textWriter, object value)
		{
			Serialize (new JsonWriter (textWriter, MaxJsonLength), value);
		}

		/// <summary>
		/// Serializes the specified <see cref="Object"/> and writes the Json structure
		/// to a <c>Stream</c> using the specified <see cref="JsonWriter"/>. 
		/// </summary>
		/// <param name="jsonWriter">The <see cref="JsonWriter"/> used to write the Json structure.</param>
		/// <param name="value">The <see cref="Object"/> to serialize.</param>
		void Serialize (JsonWriter jsonWriter, object value)
		{
			SerializeValue (jsonWriter, value);
		}

		private void SerializeValue (JsonWriter writer, object value)
		{
			//JsonConverter converter;
			_currentRecursionCounter++;
			if (RecursionLimit > 0 && _currentRecursionCounter > RecursionLimit) {
				throw new ArgumentException ("RecursionLimit exceeded.");
			}

			if (value == null) {
				writer.WriteNull ();
			} else {
				JavaScriptConverter jsconverter = _context.GetConverter (value.GetType ());
				if (jsconverter != null) {
					value = jsconverter.Serialize (value, _context);
					if (value == null) {
						writer.WriteNull ();
						return;
					}
				}

				Type valueType = value.GetType ();
				switch (Type.GetTypeCode (valueType)) {
				case TypeCode.String:
					writer.WriteValue ((string)value);
					break;
				case TypeCode.Char:
					writer.WriteValue ((char)value);
					break;
				case TypeCode.Boolean:
					writer.WriteValue ((bool)value);
					break;
				case TypeCode.SByte:
					writer.WriteValue ((sbyte)value);
					break;
				case TypeCode.Int16:
					writer.WriteValue ((short)value);
					break;
				case TypeCode.UInt16:
					writer.WriteValue ((ushort)value);
					break;
				case TypeCode.Int32:
					writer.WriteValue ((int)value);
					break;
				case TypeCode.Byte:
					writer.WriteValue ((byte)value);
					break;
				case TypeCode.UInt32:
					writer.WriteValue ((uint)value);
					break;
				case TypeCode.Int64:
					writer.WriteValue ((long)value);
					break;
				case TypeCode.UInt64:
					writer.WriteValue ((ulong)value);
					break;
				case TypeCode.Single:
					writer.WriteValue ((float)value);
					break;
				case TypeCode.Double:
					writer.WriteValue ((double)value);
					break;
				case TypeCode.DateTime:
					writer.WriteValue ((DateTime)value);
					break;
				case TypeCode.Decimal:
					writer.WriteValue ((decimal)value);
					break;
				default:
					

					ThrowOnReferenceLoop (writer, value);
					writer.SerializeStack.Push (value);
					try {
						Type genDictType;
						if (value is IDictionary)
							SerializeDictionary (writer, (IDictionary)value);
						else if (value is IDictionary<string, object>)
							SerializeDictionary (writer, (IDictionary<string, object>)value, null);
						else if ((genDictType = ReflectionUtils.GetGenericDictionary (valueType)) != null)
							SerializeDictionary (writer, new GenericDictionaryLazyDictionary (value, genDictType), null);
						else if (value is IEnumerable) {
							SerializeEnumerable (writer, (IEnumerable)value);
						} else {
							SerializeCustomObject (writer, value, valueType);
						}
					} finally {

						object x = writer.SerializeStack.Pop ();
						if (x != value)
							throw new InvalidOperationException ("Serialization stack is corrupted");
					}

					break;
				}
			}

			_currentRecursionCounter--;
		}

		private void ThrowOnReferenceLoop (JsonWriter writer, object value)
		{
			switch (_referenceLoopHandling) {
			case ReferenceLoopHandling.Error:
				if (writer.SerializeStack.Contains (value))
					throw new JsonSerializationException ("Self referencing loop");
				break;
			case ReferenceLoopHandling.Ignore:
				// return from method
				return;
			case ReferenceLoopHandling.Serialize:
				// continue
				break;
			default:
				throw new InvalidOperationException (string.Format ("Unexpected ReferenceLoopHandling value: '{0}'", _referenceLoopHandling));
			}
		}

		private void SerializeEnumerable (JsonWriter writer, IEnumerable values)
		{
			writer.WriteStartArray ();

			foreach (object value in values)
				SerializeValue (writer, value);

			writer.WriteEndArray ();
		}

		private void SerializeDictionary (JsonWriter writer, IDictionary values)
		{
			writer.WriteStartObject ();

			foreach (DictionaryEntry entry in values)
				SerializePair (writer, entry.Key.ToString (), entry.Value);

			writer.WriteEndObject ();
		}

		private void SerializeDictionary (JsonWriter writer, IDictionary<string, object> values, string typeID)
		{
			writer.WriteStartObject ();

			if (typeID != null) {
				SerializePair (writer, JavaScriptSerializer.SerializedTypeNameKey, typeID);
			}

			foreach (KeyValuePair<string, object> entry in values)
				SerializePair (writer, entry.Key, entry.Value);

			writer.WriteEndObject ();
		}

		private void SerializeCustomObject (JsonWriter writer, object value, Type valueType)
		{
			if (value is Uri) {
				Uri uri = value as Uri;
				writer.WriteValue (uri.GetComponents (UriComponents.AbsoluteUri, UriFormat.UriEscaped));
				return;
			}
			if (valueType == typeof(Guid)) {
				writer.WriteValue (((Guid)value).ToString ());
				return;
			}

			string typeID = null;
			if (_typeResolver != null) {
				typeID = _typeResolver.ResolveTypeId (valueType);
			}

			SerializeDictionary (writer, new SerializerLazyDictionary (value), typeID);
		}

		private void SerializePair (JsonWriter writer, string key, object value)
		{
			writer.WritePropertyName (key);
			SerializeValue (writer, value);
		}

		#endregion
	}
}

File: appverse-platform-android/src/com/gft/unity/android/util/json/JSONSerializer.java
/*
 Copyright (c) 2012 GFT Appverse, S.L., Sociedad Unipersonal.

 This Source  Code Form  is subject to the  terms of  the Appverse Public License 
 Version 2.0  ("APL v2.0").  If a copy of  the APL  was not  distributed with this 
 file, You can obtain one at http://appverse.org/legal/appverse-license/.

 Redistribution and use in  source and binary forms, with or without modification, 
 are permitted provided that the  conditions  of the  AppVerse Public License v2.0 
 are met.

 THIS SOFTWARE IS PROVIDED BY THE  COPYRIGHT HOLDERS  AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS  OR IMPLIED WARRANTIES, INCLUDING, BUT  NOT LIMITED TO,   THE IMPLIED
 WARRANTIES   OF  MERCHANTABILITY   AND   FITNESS   FOR A PARTICULAR  PURPOSE  ARE
 DISCLAIMED. EXCEPT IN CASE OF WILLFUL MISCONDUCT OR GROSS NEGLIGENCE, IN NO EVENT
 SHALL THE  COPYRIGHT OWNER  OR  CONTRIBUTORS  BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL,  SPECIAL,   EXEMPLARY,  OR CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT
 LIMITED TO,  PROCUREMENT OF SUBSTITUTE  GOODS OR SERVICES;  LOSS OF USE, DATA, OR
 PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) 
 ARISING  IN  ANY WAY OUT  OF THE USE  OF THIS  SOFTWARE,  EVEN  IF ADVISED OF THE 
 POSSIBILITY OF SUCH DAMAGE.
 */
package com.gft.unity.android.util.json;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import com.gft.unity.core.system.SystemLogger;
import com.gft.unity.core.system.SystemLogger.Module;

// TODO add mechanism to skip properties
public class JSONSerializer {

	private static final SystemLogger LOG = SystemLogger.getInstance();

	private static final String GETTER_PREFIX = "get";
	private static final int GETTER_PREFIX_LENGTH = GETTER_PREFIX.length();
	private static final String SETTER_PREFIX = "set";

	private static final Set<Class<?>> WRAPPER_TYPES = new HashSet<Class<?>>();

	static {
		WRAPPER_TYPES.add(Boolean.class);
		WRAPPER_TYPES.add(Character.class);
		WRAPPER_TYPES.add(Byte.class);
		WRAPPER_TYPES.add(Short.class);
		WRAPPER_TYPES.add(Integer.class);
		WRAPPER_TYPES.add(Long.class);
		WRAPPER_TYPES.add(Float.class);
		WRAPPER_TYPES.add(Double.class);
		WRAPPER_TYPES.add(Void.class);
	}

	public static String serialize(Object object) {
		String result;

		try {
			result = JSONObject.valueToString(serializeElement(object));
		} catch (Exception ex) {
			LOG.Log(Module.PLATFORM, "JSONSerializer serialize error: "
					+ object, ex);
			result = "\"null\"";
		}

		return result;
	}

	private static Object serializeElement(Object object) {
		Object element = null;

		Class<?> classType = object.getClass();
		if (classType.isArray()) {
			element = serializeArray(object);
		} else if (classType.isPrimitive() || WRAPPER_TYPES.contains(classType)
				|| object instanceof String) {
			element = object;
		} else if (classType.isEnum()) {
			element = ((Enum<?>) object).ordinal();
		} else if (object instanceof Map) {
			element = new JSONObject((Map<?,?>) object);
		} else {
			element = serializeBean(object);
		}

		return element;
	}

	private static Object serializeArray(Object object) {
		JSONArray jsonArray = new JSONArray();

		if (object.getClass().equals(byte[].class)) {
			for (byte objChild : (byte[]) object) {
				jsonArray.put(objChild);
			}
			// TODO optimize for other primitive types
		} else {
			for (Object objChild : (Object[]) object) {
				jsonArray.put(serializeElement(objChild));
			}
		}

		return jsonArray;
	}

	private static JSONObject serializeBean(Object object) {

		HashMap<Object, Object> jsonMap = new HashMap<Object, Object>();
		Method[] methods = object.getClass().getMethods();
		for (Method method : methods) {
			if (isPropertyGetter(object, method)) {
				String methodName = method.getName();
				String fieldName = methodName.substring(GETTER_PREFIX_LENGTH);
				Object fieldValue = null;
				try {
					Object objValue = method.invoke(object, new Object[] {});
					if (objValue == null) {
						fieldValue = JSONObject.NULL;
					} else {
						fieldValue = serializeElement(objValue);
					}
				} catch (Exception e) {
					LOG.Log(Module.PLATFORM,
							"JSONSerializer error invoking method ["
									+ methodName + "] in the object [" + object
									+ "]", e);
				}
				jsonMap.put(fieldName, fieldValue);
			}
		}

		return new JSONObject(jsonMap);
	}

	private static boolean isPropertyGetter(Object object, Method method) {

		String methodName = method.getName();
		Class<?> declaringClass = method.getDeclaringClass();

		// method must have only one parameter.
		if (method.getParameterTypes().length > 0) {
			return false;
		}

		// method must start with 'get' or 'Get' and the first character after
		// the prefix must in upper case.
		if (!methodName.toUpperCase().startsWith(GETTER_PREFIX.toUpperCase())
				|| !Character.isUpperCase(methodName
						.charAt(GETTER_PREFIX_LENGTH))) {
			return false;
		}

		// method must be declared in an Unity sub package.
		if (!declaringClass.getPackage().getName().startsWith("com.gft.unity")) {
			return false;
		}

		// method must be declared in the class or immediate superclass
		if (!declaringClass.equals(object.getClass())
				&& !declaringClass.equals(object.getClass().getSuperclass())) {
			return false;
		}

		return true;
	}

	public static Object deserialize(Class<?> classType, Object object)
			throws Exception {
		return deserializeElement(classType, object);
	}

	private static Object deserializeElement(Class<?> classType, Object object)
			throws Exception {
		Object element;

		if (classType.isArray()) {
			element = deserializeArray(classType, (JSONArray) object);
		} else if (classType.isEnum()) {
				element = deserializeEnum(classType, object);
		} else if (classType.isPrimitive() 
				|| object instanceof String
				|| object instanceof Integer
				|| object instanceof Long
				|| object instanceof Short
				|| object instanceof Double
				|| object instanceof Float) {
			element = deserializeSimple(classType, object);
		} else {
			if (JSONObject.NULL.equals(object)) {
				element = null;
			} else {
				element = deserializeBean(classType, (JSONObject) object);
			}
		}

		return element;
	}

	private static Object deserializeArray(Class<?> classType,
			JSONArray jsonArray) throws Exception {
		Object array = null;

		Class<?> elementClassType = classType.getComponentType();

		if (elementClassType.isPrimitive()) {

			if (elementClassType.equals(int.class)) {
				int[] ints = new int[jsonArray.length()];
				for (int i = 0; i < ints.length; i++) {
					ints[i] = Integer.valueOf(jsonArray.getString(i));
				}
				array = ints;
			} else if (elementClassType.equals(boolean.class)) {
				boolean[] booleans = new boolean[jsonArray.length()];
				for (int i = 0; i < booleans.length; i++) {
					booleans[i] = Boolean.valueOf(jsonArray.getString(i));
				}
				array = booleans;
			} else if (elementClassType.equals(char.class)) {
				char[] chars = new char[jsonArray.length()];
				for (int i = 0; i < chars.length; i++) {
					chars[i] = Character.valueOf(jsonArray.getString(i).charAt(
							0));
				}
				array = chars;
			} else if (elementClassType.equals(byte.class)) {
				byte[] bytes = new byte[jsonArray.length()];
				for (int i = 0; i < bytes.length; i++) {
					String jsonStringByte = jsonArray.getString(i);
					// [MOBPLAT-184] - special characters such as accented, were overfloading the Byte.valueOf() method 
					// that only accepts -128 to 127 integers.
					// To avoid this, we should first parse the string to Integer, and then cast to byte
					// UTF-8 compatible
					bytes[i] = (byte) Integer.parseInt(jsonStringByte);
				}
				array = bytes;
			} else if (elementClassType.equals(short.class)) {
				short[] shorts = new short[jsonArray.length()];
				for (int i = 0; i < shorts.length; i++) {
					shorts[i] = Short.valueOf(jsonArray.getString(i));
				}
				array = shorts;
			} else if (elementClassType.equals(long.class)) {
				long[] longs = new long[jsonArray.length()];
				for (int i = 0; i < longs.length; i++) {
					longs[i] = Long.valueOf(jsonArray.getString(i));
				}
				array = longs;
			} else if (elementClassType.equals(float.class)) {
				float[] floats = new float[jsonArray.length()];
				for (int i = 0; i < floats.length; i++) {
					floats[i] = Float.valueOf(jsonArray.getString(i));
				}
				array = floats;
			} else if (elementClassType.equals(double.class)) {
				double[] doubles = new double[jsonArray.length()];
				for (int i = 0; i < doubles.length; i++) {
					doubles[i] = Double.valueOf(jsonArray.getString(i));
				}
				array = doubles;
			}
		} else {

			Object[] objects = (Object[]) Array.newInstance(elementClassType,
					jsonArray.length());
			for (int i = 0; i < jsonArray.length(); i++) {
				objects[i] = deserializeElement(elementClassType,
						jsonArray.get(i));
			}

			array = objects;
		}

		return array;
	}

	private static Object deserializeSimple(Class<?> classType, Object object) {
		Object simple = null;

		if (classType.equals(int.class)) {
			simple = Integer.valueOf(object.toString());
		} else if (classType.equals(boolean.class)) {
			simple = Boolean.valueOf(object.toString());
		} else if (classType.equals(char.class)) {
			simple = Character.valueOf(object.toString().charAt(0));
		} else if (classType.equals(byte.class)) {
			simple = Byte.valueOf(object.toString());
		} else if (classType.equals(short.class)) {
			simple = Short.valueOf(object.toString());
		} else if (classType.equals(long.class)) {
			simple = Long.valueOf(object.toString());
		} else if (classType.equals(float.class)) {
			simple = Float.valueOf(object.toString());
		} else if (classType.equals(double.class)) {
			simple = Double.valueOf(object.toString());
		} else {
			simple = object.toString();
		}

		return simple;
	}

	private static Object deserializeEnum(Class<?> classType, Object object) {

		Object[] enums = classType.getEnumConstants();
		return enums[Integer.valueOf(object.toString())];
	}

	private static Object deserializeBean(Class<?> classType, JSONObject object)
			throws Exception {

		Object bean = classType.newInstance();

		@SuppressWarnings("unchecked")
		Iterator<String> keys = object.keys();
		Method[] methods = classType.getMethods();

		while (keys.hasNext()) {
			String key = keys.next();
			Method setter = null;
			for (Method method : methods) {
				if (method.getName().toUpperCase()
						.equals((SETTER_PREFIX + key).toUpperCase())
						&& method.getParameterTypes().length == 1) {
					setter = method;
					break;
				}
			}
			if (setter != null) {
				Object parameter = deserializeElement(
						setter.getParameterTypes()[0], object.get(key));
				setter.invoke(bean, parameter);

			}
		}

		return bean;
	}
}
